v2/pkg/engine/jsonschema/variables_schema.go

```diff
@@ -0,0 +1,461 @@
+package jsonschema
+
+import (
+	"fmt"
+
+	"github.com/wundergraph/graphql-go-tools/v2/pkg/ast"
+	"github.com/wundergraph/graphql-go-tools/v2/pkg/operationreport"
+)
+
+// VariablesSchemaBuilder creates a unified JSON schema for the variables of a GraphQL operation
+type VariablesSchemaBuilder struct {
+	operationDocument  *ast.Document
+	definitionDocument *ast.Document
+	schema             *JsonSchema
+	report             *operationreport.Report
+	// Track recursion depth for each type to handle recursive types
+	recursionTracker  map[string]int
+	maxRecursionDepth int
+}
+
+// NewVariablesSchemaBuilder creates a new VariablesSchemaBuilder with default settings
+func NewVariablesSchemaBuilder(operationDocument, definitionDocument *ast.Document) *VariablesSchemaBuilder {
+	return NewVariablesSchemaBuilderWithOptions(operationDocument, definitionDocument, 3)
+}
+
+// NewVariablesSchemaBuilderWithOptions creates a new VariablesSchemaBuilder with custom options
+func NewVariablesSchemaBuilderWithOptions(operationDocument, definitionDocument *ast.Document, maxRecursionDepth int) *VariablesSchemaBuilder {
+	return &VariablesSchemaBuilder{
+		operationDocument:  operationDocument,
+		definitionDocument: definitionDocument,
+		schema:             NewObjectSchema(),
+		report:             &operationreport.Report{},
+		recursionTracker:   make(map[string]int),
+		maxRecursionDepth:  maxRecursionDepth,
+	}
+}
+
+// Build traverses the operation and builds a unified JSON schema for its variables
+func (v *VariablesSchemaBuilder) Build() (*JsonSchema, error) {
+	v.schema = NewObjectSchema()
+	v.recursionTracker = make(map[string]int) // Reset recursion tracker for each build
+
+	// Extract descriptions from root fields
+	var descriptions []string
+	operationDefinition := v.operationDocument.OperationDefinitions[0]
```

Comment:
What if there is no operation definition?

I recommend to look through RootNodes of operationDocument to check if there is a document. How will you deal with documents containing multiple operations? fail?

Reply:
Adressed